[de-duper] Too many links

JonWBedard opened the pull request:
#### dbef21b6b3541049e000ee180ae34ec600896d89
<pre>
[de-duper] Too many links
<a href="https://bugs.webkit.org/show_bug.cgi?id=282471">https://bugs.webkit.org/show_bug.cgi?id=282471</a>
<a href="https://rdar.apple.com/139101004">rdar://139101004</a>

Reviewed by Sam Sneddon.

* Tools/CISupport/Shared/de-duper:
(main): Don&apos;t try to link the storage directory to itself, and gracefully handle the maximum
number of file links.

Canonical link: <a href="https://commits.webkit.org/286617@main">https://commits.webkit.org/286617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c8b51c2b338599ebf345931f29fdce92cf58c61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76537 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/55572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/81064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/27813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78654 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/64714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/3865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/81064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/27813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79604 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/64714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/81064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/64714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/26137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/64714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/82513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/3913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/3865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/82513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/76214 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/4066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/82513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11841 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/3860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3883 "Build is in progress. Recent messages:") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/7313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/5641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -27,6 +27,7 @@ import os
 import sys
 
 CHUNK_SIZE = 8192
+MAX_LINKS = 64990
```
Given this varies between filesystems, I'd rather we didn't hardcode this (and, e.g., APFS (or maybe Darwin generally) seems to have a lower limit, and NTFS has an even lower limit).

`os.pathconf(path, "PC_LINK_MAX")` should give the number of links `path` can have to it, on all POSIX-like systems (and we probably don't care about non-POSIX-like systems here).

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -96,16 +97,20 @@ def main(args):
 
     for root, _, files in os.walk(args.root, topdown=False):
+        if root == args.storage:
+            continue
```
This seems like a noteworthy change, but also we probably want `os.path.samefile` here, rather than relying on the paths being identical.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,21 +119,28 @@ def main(args):
             sha_value = sha1(full_path)
+            sha_path = os.path.join(args.storage, f'{sha_value}.png')
+            if sha_path == full_path:
+                continue
```
This should be unreachable with the check at the top of the `os.walk` loop, right? Can we just make this an assert instead, given this should never happen?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,21 +119,28 @@ def main(args):
+                    existing_links[sha_value] += 1
+                    if existing_links[sha_value] > MAX_LINKS:
+                        os.remove(sha_path)
+                        del existing_links[sha_value]
```
This behaviour seems quite subtle and probably worth commenting: we're effectively allowing a new file to be created in storage once we reach `MAX_LINKS`, but this doesn't impact the existing files linked to the file we're removing from this path.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -96,16 +97,20 @@ def main(args):
                 cleaned_up += 1
             else:
-                existing_links.add(sha_value)
+                existing_links[sha_value] = connected_to
```
The caching of the inode count makes me slightly nervous, because while we probably don't expect external processes to mutate it, there is always that risk.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,21 +119,28 @@ def main(args):
-                    os.link(os.path.join(args.storage, f'{sha_value}.png'), full_path)
+                    os.link(sha_path, full_path)
+                    existing_links[sha_value] += 1
+                    if existing_links[sha_value] > MAX_LINKS:
```
Again, the caching here makes me slightly nervous, because what if we have a lot more links being created by some other process, we could still have `os.link` raise above?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,21 +119,28 @@ def main(args):
                 else:
                     os.remove(full_path)
-                    os.link(os.path.join(args.storage, f'{sha_value}.png'), full_path)
+                    os.link(sha_path, full_path)
```
I do still wonder if we want to do something like:

```
try:
    os.link(sha_path, full_path)
except OSError as e:
    if e.errno == errno.EMLINK:
        os.remove(sha_path)
        os.link(sha_path, full_path)
    else:
        raise
```

To guard against other processes also creating links?

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,21 +119,28 @@ def main(args):
-                    os.link(full_path, os.path.join(args.storage, f'{sha_value}.png'))
-                existing_links.add(sha_value)
+                    os.link(full_path, sha_path)
+                existing_links[sha_value] = 2
```
I wonder if there's any world in which we get a different `inode` count… It would seem safer to read from the disk, again, especially rather than hardcoding this value?

JonWBedard replied:
I was trying to avoid reading from disk again, but reading the rest of your comments, this feels like the correct design. The updated PR is removing the `existing_links` variable entirely in favor of reading from disk.

gsnedders reviewed (changes requested):
tl;dr:

The caching makes me slightly nervous (what if another process — e.g. a backup one — is also increment inode count at the same time?), and the hardcoding of magic values doesn't seem great.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,23 +120,31 @@ def main(args):
+                elif os.stat(sha_path).st_nlink >= max_links - 1:
+                    # If we've exceeded the maximum link count, remove the SHA file from the hard-link network
+                    # and create a new network. In practice, the previous network will slowly be deleted
+                    # as old results are pruned
```
"hard-link network" shows up nothing on Google as a phrase; you simply have a file with multiple directory entries if you believe most sources. Maybe we should just go with something like:

> If we're near the maximum link count, remove the directory entry within storage pointing at that file, and create a new directory entry within storage pointing at another (identical) file.  In practice, the previous file will slowly be deleted as old results are pruned.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -86,7 +86,9 @@ def main(args):
     processed = 0
 
-    existing_links = set()
+    max_links = os.pathconf(args.storage, 'PC_LINK_MAX')
```
(A probably theoretical nit: hypothetically a compliant POSIX implementation could have different values of `max_links` for each file. That said, see my comment below which would imply we don't need this at all.)

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,23 +120,31 @@ def main(args):
+            if os.path.isfile(sha_path):
                 if args.dry_run:
                     print(f'[dry-run] Linking {full_path} to {sha_value}.png')
+                elif os.stat(sha_path).st_nlink >= max_links - 1:
```
I wonder what the cost of all these checks actually is. Quite possibly not insignificant, given we're calling `stat` each time?

I do really still wonder if something like https://github.com/WebKit/WebKit/pull/36071/files#r1828491761 is actually the right option here, just dealing with the error from `link`, and just use that as the `not args.dry_run` case? Specifically, I believe this has th same semantics as the `elif/else` branches here:

```python
os.remove(full_path)
try:
    os.link(sha_path, full_path)
except OSError as e:
    if e.errno == errno.EMLINK:
        # sha_path already has LINK_MAX links, so create a copy of that file at
        # full_path, and create a new hard link within storage
        # pointing at another (identical) file. In practice, the previous file will
        # slowly be deleted as old results are pruned.
        shutil.copyfile(sha_path, full_path)
        os.remove(sha_path)
        os.link(full_path, sha_path)
    else:
        raise
```

JonWBedard replied:
Gave this implementation a try, when combined with checking for file existence, it works quite well.

I had originally tried this. But it doesn't work well when paired with the old code which attempts to keep track of which images are in storage in program memory instead of checking disk each time. When we completely abandon in-memory representations and rely on the filesystem entirely, this works.

gsnedders reviewed (commented):
I still think relying on `OSError` being raised is probably better than calling `stat`, as it both avoids the cost of the `stat` call and means we never have a race with any other process.

Also looking at this again, I'm kinda eying all the `os.remove` calls somewhat nervously. It makes me wonder if it's worthwhile writing to a temporary file and then using `os.replace` to replace the existing file, thus always preserving data? (i.e., an exception being thrown after the `os.remove` but before we create the new link wouldn't be an issue)

Of course, taken to extreme we still don't have that guarantee — the system could panic, and the filesystem could be in an inconsistent state, etc., but at least then we'll never have a syscall failing causing data loss.

gsnedders commented on Tools/CISupport/Shared/de-duper:
```diff
@@ -114,23 +119,37 @@ def main(args):
+                            # full_path, and create a new hard link within storage pointing at another
+                            # identical file. In practice, the previous network of hard-linked files will
+                            # slowly be deleted as old results are pruned.
+                            os.remove(sha_path)
+                            os.link(full_path, sha_path)
```
Nit: we should maybe call `os.replace` here, or comment why not.

(I think in practice, because `sha_path` is never actually directly referenced, that it simply doesn't matter if we end up in a state where that file is deleted but not recreated?)

JonWBedard replied:
I'm make the comment!

Indeed, the answer is that we just don't care if the file is deleted by not recreated.

JonWBedard committed:
[de-duper] Too many links
https://bugs.webkit.org/show_bug.cgi?id=282471
rdar://139101004

Reviewed by Sam Sneddon.

* Tools/CISupport/Shared/de-duper:
(main): Don't try to link the storage directory to itself, and gracefully handle the maximum
number of file links.

Canonical link: https://commits.webkit.org/286617@main
